File: leetcode/easy/94-inorder/stack.cpp
/**
 * @file stack.cpp
 * @author Vortexx2
 * @brief Problem 94 - Binary Tree Inorder Traversal
 * @date 21-07-2021
 * 
 * Runtime - 0 ms 
 * Memory Usage - 8.5 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<int> inorderTraversal(TreeNode *root)
  {
    stack<TreeNode *> st;
    vector<int> res;

    while (true)
    {
      while (root)
      {
        st.push(root);
        root = root->left;
      }

      if (st.empty())
        break;

      root = st.top();
      st.pop();
      res.push_back(root->val);
      root = root->right;
    }

    return res;
  }
};

File: leetcode/easy/94-inorder/stack.py
"""
  @author Vortexx2
  Problem 94 - Binary Tree Inorder Traversal
  
  Runtime - 32 ms
  Memory Usage - 14.3 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def inorderTraversal(self, root: TreeNode) -> List[int]:
    res = []
    st = []

    while True:

      while root:
        st.append(root)
        root = root.left

      if not st:
        break

      root = st.pop()
      res.append(root.val)
      root = root.right

    return res
